### PR TITLE
Complete libieeep1788_class.jl

### PR DESCRIPTION
**Pull request by krish8484**

### CODE REVIEW

**Comment by krish8484** [comment]
`interval(Inf, Inf)`, `interval(-Inf, Inf)` and Interval(NaN, NaN) are returning argument errors now (according to the tests)

File: test/interval_tests/construction.jl [hunk 1]
```diff
@@ -64,7 +64,7 @@ const eeuler = Base.MathConstants.e
     # a < Inf and b > -Inf
     @test @interval(1e300) == Interval(9.999999999999999e299, 1.0e300)
     @test @interval(-1e307) == Interval(-1.0000000000000001e307, -1.0e307)
-    @test @interval(Inf) == IntervalArithmetic.wideinterval(Inf)
```
Comment (dpsanders) [inline test/interval_tests/construction.jl:67]: I think we would still want `@interval(Inf)` to return `interval(prevfloat(Inf), Inf)`

And similarly for `-Inf`.

`@interval` is allowed to do special things.

Comment (krish8484) [reply test/interval_tests/construction.jl:67]: okay, so should we comment out tests like these `@test @interval("[+infinity]") == ∅`?

Comment (dpsanders) [reply test/interval_tests/construction.jl:67]: If we are deliberately changing the behaviour then we should change the tests to test the new behaviour.

Comment (lbenet) [reply test/interval_tests/construction.jl:67]: > I think we would still want `@interval(Inf)` to return `interval(prevfloat(Inf), Inf)`
> 
> And similarly for `-Inf`.
> 
> `@interval` is allowed to do special things.

Same opinion here. Also, I think that is the expected behavior according to the standard.

**Comment by krish8484** [comment]
`@decorated` responds to `setprecision` now
Master
```
julia> typeof(@decorated(1, 2))
DecoratedInterval{Float64}

julia> setprecision(Interval, 256)
256

julia> typeof(@decorated(1, 2))
DecoratedInterval{Float64}
```
Branch
```
julia> typeof(@decorated(1, 2))
DecoratedInterval{Float64}

julia> setprecision(Interval, 256)
256

julia> typeof(@decorated(1, 2))
DecoratedInterval{BigFloat}
```
`@decorated` now throws an `ArgumentError` just like `@interval` for improper inputs

Branch
```
julia> @decorated(2, -1)
ERROR: ArgumentError: `[2.0, -1.0]` is not a valid interval. Need `a ≤ b` to construct `interval(a, b)`.
Stacktrace:
 [1] interval(::Float64, ::Float64) at C:\Users\ASUS\.julia\dev\IntervalArithmetic\src\intervals\intervals.jl:107
 [2] top-level scope at REPL[2]:1

julia> @decorated("[  -1.0  , 1.0]_fooo") #this was one of ITF1788 Test
ERROR: LoadError: ArgumentError: Cannot process fooo as decoration
Stacktrace:
 [1] parse(::Type{DecoratedInterval{Float64}}, ::String) at C:\Users\ASUS\.julia\dev\IntervalArithmetic\src\parsing.jl:48
 [2] @decorated(::LineNumberNode, ::Module, ::Vararg{Any,N} where N) at C:\Users\ASUS\.julia\dev\IntervalArithmetic\src\decorations\intervals.jl:138
in expression starting at REPL[3]:1
```
Better string handling for `@interval` and `@decorated`

Branch
```

julia> @decorated("[ Nai  ]") == nai()
true

julia> @decorated("[-1.0, +infinity]_def") == DecoratedInterval(Interval(-1.0, Inf), def)
true

julia> @test_throws ArgumentError @interval("[ -inf ,  INF ]_com")
Test Passed
      Thrown: ArgumentError
```

However, these are some tests that are still failing, any suggestions on how can we solve them?
```
julia>  @interval("[1.0E+400 ]") == @interval(0x1.fffffffffffffp+1023, Inf)
ERROR: ArgumentError: Unable to process string [1.0E+400 ] as interval
Stacktrace:
 [1] parse(::Type{Interval{Float64}}, ::String) at C:\Users\ASUS\.julia\dev\IntervalArithmetic\src\parsing.jl:297
 [2] atomic(::Type{Interval{Float64}}, ::String) at C:\Users\ASUS\.julia\dev\IntervalArithmetic\src\intervals\conversion.jl:75
 [3] top-level scope at REPL[30]:1
caused by [exception 1]
Base.Meta.ParseError("overflow in numeric constant \"1.0e+400\"")

julia> @interval("10?3e380") == @interval(0x1.fffffffffffffp+1023, Inf)
ERROR: ArgumentError: cannot parse "7.0e380" as Float64
Stacktrace:
 [1] _parse_failure(::Type, ::String, ::Int64, ::Int64) at .\parse.jl:370 (repeats 2 times)
 [2] #tryparse_internal#348 at .\parse.jl:366 [inlined]
 [3] tryparse_internal at .\parse.jl:364 [inlined]
 [4] #parse#349 at .\parse.jl:376 [inlined]
 [5] parse at .\parse.jl:376 [inlined]
 [6] parse(::Type{Interval{Float64}}, ::String) at C:\Users\ASUS\.julia\dev\IntervalArithmetic\src\parsing.jl:212
 [7] atomic(::Type{Interval{Float64}}, ::String) at C:\Users\ASUS\.julia\dev\IntervalArithmetic\src\intervals\conversion.jl:75
 [8] top-level scope at REPL[31]:1

julia> @decorated("10?1800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000_com") == DecoratedInterval(Interval(-Inf, Inf), dac)
ERROR: LoadError: ArgumentError: cannot parse "1800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e-0" as Float64
Stacktrace:
 [1] _parse_failure(::Type, ::String, ::Int64, ::Int64) at .\parse.jl:370 (repeats 2 times)
 [2] #tryparse_internal#348 at .\parse.jl:366 [inlined]
 [3] tryparse_internal at .\parse.jl:364 [inlined]
 [4] #parse#349 at .\parse.jl:376 [inlined]
 [5] parse at .\parse.jl:376 [inlined]
 [6] parse(::Type{Interval{Float64}}, ::SubString{String}) at C:\Users\ASUS\.julia\dev\IntervalArithmetic\src\parsing.jl:119
 [7] parse(::Type{DecoratedInterval{Float64}}, ::String) at C:\Users\ASUS\.julia\dev\IntervalArithmetic\src\parsing.jl:42
 [8] @decorated(::LineNumberNode, ::Module, ::Vararg{Any,N} where N) at C:\Users\ASUS\.julia\dev\IntervalArithmetic\src\decorations\intervals.jl:138
in expression starting at REPL[32]:1

julia>  @decorated("[ -inf ,  INF ]_com") == nai()
false

julia>  @decorated("[ -inf ,  INF ]_com")
DecoratedInterval(Interval(-Inf, Inf), com)
```
There are few more which more failing because of strict rounding

**Comment by krish8484** [comment]
There are multiple tests in this file like
```
@interval("10?3e380") == @interval(0x1.fffffffffffffp+1023, Inf)
```
These tests expect `Interval(floatmax(), Inf)` as output when the input(like in the above case) exceeds `floatmax()`. This is not ideal since 
```
julia> floatmax()
1.7976931348623157e308
```
which is very less than the desired input which is `7e380` and `13e380`.

How should we handle this in `IntervalArithmetic.jl`?
@dpsanders @lbenet

**Comment by dpsanders** [comment]
As you say the maximum representan le float is around 1e308 so that is the correct result. There's nothing else that can be done in the realm of Float64.

File: src/parsing.jl [hunk 2]
```diff
@@ -211,56 +234,81 @@ function parse(::Type{Interval{T}}, s::AbstractString) where T
...
+            m.captures[1] = strip(m.captures[1], [' '])
+            m.captures[2] = strip(m.captures[2], [' '])
+            lo, hi = m.captures
 
-    if m == nothing
+            m.captures[1] = lowercase(m.captures[1])
+            m.captures[2] = lowercase(m.captures[2])
 
-        m = match(r"\[(.*)\]", s)  # string like "[1]"
+            if m.captures[2] == "+infinity" || m.captures[2] == "" || m.captures[2] == "inf" || m.captures[2] == "+inf" || m.captures[2] == "infinity"
```
Comment (dpsanders) [inline src/parsing.jl:256]: You can use `m.captures[2] in ("+infinity", "inf", ...)` etc.

**Comment by krish8484** [comment]
All the tests from this file except two are passing now. Those 2 tests are failing because it is related to issue #395.

**Pull request closed without merging**